File: dotnet/GHIElectronics.DUELink/CoProcessor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {
        public class CoProcessorController {
            SerialInterface serialPort;
            StreamController stream;
            public CoProcessorController(SerialInterface serialPort, StreamController stream) {
                this.serialPort = serialPort;
                this.stream = stream;
            }


            public bool Erase() {

                // we can't check response as Asio(1) there will be no response                                

                this.serialPort.WriteCommand("CoprocE()");

                var ret = this.serialPort.ReadResponse();

                return ret.success;
            }

            public bool Program() {

                // need Xmodem 1K, TODO
                // 
                throw new NotImplementedException(); ;
            }

            public bool Reset() {
                this.serialPort.WriteCommand("CoprocS()");

                var ret = this.serialPort.ReadResponse();

                return ret.success;
            }

            public string Version() {
                this.serialPort.WriteCommand("CoprocV()");

                var ret = this.serialPort.ReadResponse();

                return ret.response;
            }

            public int Write(byte[] data) {

                //var write_array = string.Empty;

                //write_array = "[";

                //for (var i = 0; i < data.Length; i++) {
                //    write_array += data[i];

                //    if (i < data.Length - 1)
                //        write_array += ",";
                //}

                //write_array += "]";

                var count = data.Length;

                // declare b9 array
                var cmd = string.Format("dim b9[{0}])", count);

                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                // write data to b9
                var written = this.stream.WriteBytes("b9", data);

                // write b9 to co-pro
                this.serialPort.WriteCommand("CoprocW(b9)");
                var ret = this.serialPort.ReadResponse();

                if (ret.success)
                    return written;

                return 0;
            }

            public int Read(byte[] data) {
                var count = data.Length;
                
                var cmd = string.Format("dim b9[{0}])", count);

                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                this.serialPort.WriteCommand("CoprocR(b9)");

                var ret = this.serialPort.ReadResponse();

                // use stream to read b9
                var read = this.stream.ReadBytes("b9", data);

                if (ret.success)
                    return read;

                return 0;
            }

        }
    }
}

File: micropython/standard-library/duelink/coprocessor.py

class CoProcessorController:    

    def __init__(self, serialPort,stream):
        self.transport = serialPort
        self.stream = stream

    def Erase(self)->bool:
        cmd = "CoprocE()"
        self.transport.WriteCommand(cmd)
        r,s = self.transport.ReadResponse()
        return s
    
    def Program(self)->bool:
        # need Xmodem 1K, TODO
        raise Exception("Not implemented")
            
    def Reset(self)->bool:
        cmd = "CoprocS()"
        self.transport.WriteCommand(cmd)
        r,s = self.transport.ReadResponse()
        return r

    def Version(self) -> str:
        cmd = "CoprocV()"
        self.transport.WriteCommand(cmd)
        r,s = self.transport.ReadResponse()

        return (s)
    
    def Write(self, dataWrite: bytes) -> int:
        count = len(dataWrite)

        # declare b9 array
        cmd = f"dim b9[{count}]"
        self.transport.WriteCommand(cmd)
        self.transport.ReadResponse()

        # write data to b9
        written = self.stream.WriteBytes("b9",dataWrite)

        # write b9 to co-pro
        self.transport.WriteCommand("CoprocW(b9)")
        r,s = self.transport.ReadResponse()
        
        if written == len(dataWrite):            
            return written
        return 0        
    
    def Read(self, dataRead: bytes)-> int:
        count = len(dataRead)

        # declare b9 array
        cmd = f"dim b9[{count}]"
        self.transport.WriteCommand(cmd)
        self.transport.ReadResponse()

        # read data to b9
        self.transport.WriteCommand("CoprocR(b9)")
        r,s = self.transport.ReadResponse()

        # read b9 by stream
        read = self.stream.ReadBytes("b9",dataRead )

        if r:
            return read
        return 0 
